File: vb.net/SoundCapture/SoundCaptureBase.vb
Imports System.Threading
Imports Microsoft.DirectX.DirectSound
Imports Microsoft.Win32.SafeHandles

''' <summary>
''' Base class to capture audio samples.
''' </summary>
Public MustInherit Class SoundCaptureBase
	Implements IDisposable
	Const BufferSeconds As Integer = 3
	Const NotifyPointsInSecond As Integer = 2

	' change in next two will require also code change
	Const BitsPerSample As Integer = 16
	Const ChannelCount As Integer = 1

	Private m_sampleRate As Integer = 44100
	Private m_isCapturing As Boolean = False
	Private disposed As Boolean = False

	Public ReadOnly Property IsCapturing() As Boolean
		Get
			Return m_isCapturing
		End Get
	End Property

	Public Property SampleRate() As Integer
		Get
			Return m_sampleRate
		End Get
		Set
			If m_sampleRate <= 0 Then
				Throw New ArgumentOutOfRangeException()
			End If

			EnsureIdle()

			m_sampleRate = value
		End Set
	End Property

	Private capture As Capture
	Private buffer As CaptureBuffer
	Private notify As Notify
	Private bufferLength As Integer
	Private positionEvent As AutoResetEvent
	Private positionEventHandle As SafeWaitHandle
	Private terminated As ManualResetEvent
	Private thread As Thread
	Private device As SoundCaptureDevice

	Public Sub New()

		Me.New(SoundCaptureDevice.GetDefaultDevice())
	End Sub

	Public Sub New(device As SoundCaptureDevice)
		Me.device = device

		positionEvent = New AutoResetEvent(False)
		positionEventHandle = positionEvent.SafeWaitHandle
		terminated = New ManualResetEvent(True)
	End Sub

	Private Sub EnsureIdle()
		If IsCapturing Then
			Throw New SoundCaptureException("Capture is in process")
		End If
	End Sub

	''' <summary>
	''' Starts capture process.
	''' </summary>
	Public Sub Start()
		EnsureIdle()

		m_isCapturing = True

		Dim format As New WaveFormat()
		format.Channels = ChannelCount
		format.BitsPerSample = BitsPerSample
		format.SamplesPerSecond = SampleRate
		format.FormatTag = WaveFormatTag.Pcm
		format.BlockAlign = CShort((format.Channels * format.BitsPerSample + 7) \ 8)
		format.AverageBytesPerSecond = format.BlockAlign * format.SamplesPerSecond

		bufferLength = format.AverageBytesPerSecond * BufferSeconds
		Dim desciption As New CaptureBufferDescription()
		desciption.Format = format
		desciption.BufferBytes = bufferLength

		capture = New Capture(device.Id)
		buffer = New CaptureBuffer(desciption, capture)

		Dim waitHandleCount As Integer = BufferSeconds * NotifyPointsInSecond
		Dim positions As BufferPositionNotify() = New BufferPositionNotify(waitHandleCount - 1) {}
		For i As Integer = 0 To waitHandleCount - 1
			Dim position As New BufferPositionNotify()
			position.Offset = (i + 1) * bufferLength \ positions.Length - 1
			position.EventNotifyHandle = positionEventHandle.DangerousGetHandle()
			positions(i) = position
		Next

		notify = New Notify(buffer)
		notify.SetNotificationPositions(positions)

		terminated.Reset()
		thread = New Thread(New ThreadStart(AddressOf ThreadLoop))
		thread.Name = "Sound capture"
		thread.Start()
	End Sub

	Private Sub ThreadLoop()
		buffer.Start(True)
		Try
			Dim nextCapturePosition As Integer = 0
			Dim [handles] As WaitHandle() = New WaitHandle() {terminated, positionEvent}
			While WaitHandle.WaitAny([handles]) > 0
				Dim capturePosition As Integer, readPosition As Integer
				buffer.GetCurrentPosition(capturePosition, readPosition)

				Dim lockSize As Integer = readPosition - nextCapturePosition
				If lockSize < 0 Then
					lockSize += bufferLength
				End If
				If (lockSize And 1) <> 0 Then
					lockSize -= 1
				End If

				Dim itemsCount As Integer = lockSize >> 1

				Dim data As Short() = DirectCast(buffer.Read(nextCapturePosition, GetType(Short), LockFlag.None, itemsCount), Short())
				ProcessData(data)
				nextCapturePosition = (nextCapturePosition + lockSize) Mod bufferLength
			End While
		Finally
			buffer.[Stop]()
		End Try
	End Sub

	''' <summary>
	''' Processes the captured data.
	''' </summary>
	''' <param name="data">Captured data</param>
	Protected MustOverride Sub ProcessData(data As Short())

	''' <summary>
	''' Stops capture process.
	''' </summary>
	Public Sub [Stop]()
		If m_isCapturing Then
			m_isCapturing = False

			terminated.[Set]()
			thread.Join()

			notify.Dispose()
			buffer.Dispose()
			capture.Dispose()
		End If
	End Sub

	Private Sub IDisposable_Dispose() Implements IDisposable.Dispose
		Dispose(True)
	End Sub

	Protected Overrides Sub Finalize()
		Try
			Dispose(False)
		Finally
			MyBase.Finalize()
		End Try
	End Sub

	Private Sub Dispose(disposing As Boolean)
		If disposed Then
			Return
		End If

		disposed = True
		GC.SuppressFinalize(Me)
		If IsCapturing Then
			[Stop]()
		End If
		positionEventHandle.Dispose()
		positionEvent.Close()
		terminated.Close()
	End Sub
End Class

File: trunk/SoundCapture/SoundCaptureBase.cs
using System;
using System.Threading;
using Microsoft.DirectX.DirectSound;
using Microsoft.Win32.SafeHandles;

namespace SoundCapture
{
    /// <summary>
    /// Base class to capture audio samples.
    /// </summary>
    public abstract class SoundCaptureBase : IDisposable
    {
        const int BufferSeconds = 3;
        const int NotifyPointsInSecond = 2;

        // change in next two will require also code change
        const int BitsPerSample = 16; 
        const int ChannelCount = 1; 

        int sampleRate = 44100;
        bool isCapturing = false;
        bool disposed = false;

        public bool IsCapturing
        {
            get { return isCapturing; }
        }

        public int SampleRate
        {
            get { return sampleRate; }
            set 
            {
                if (sampleRate <= 0) throw new ArgumentOutOfRangeException();

                EnsureIdle();

                sampleRate = value; 
            }
        }
        
        Capture capture;
        CaptureBuffer buffer;
        Notify notify;
        int bufferLength;
        AutoResetEvent positionEvent;
        SafeWaitHandle positionEventHandle;
        ManualResetEvent terminated;
        Thread thread;
        SoundCaptureDevice device;

        public SoundCaptureBase()
            : this(SoundCaptureDevice.GetDefaultDevice())
        {

        }

        public SoundCaptureBase(SoundCaptureDevice device)
        {
            this.device = device;

            positionEvent = new AutoResetEvent(false);
            positionEventHandle = positionEvent.SafeWaitHandle;
            terminated = new ManualResetEvent(true);
        }

        private void EnsureIdle()
        {
            if (IsCapturing)
                throw new SoundCaptureException("Capture is in process");
        }

        /// <summary>
        /// Starts capture process.
        /// </summary>
        public void Start()
        {
            EnsureIdle();

            isCapturing = true;

            WaveFormat format = new WaveFormat();
            format.Channels = ChannelCount;
            format.BitsPerSample = BitsPerSample;
            format.SamplesPerSecond = SampleRate;
            format.FormatTag = WaveFormatTag.Pcm;
            format.BlockAlign = (short)((format.Channels * format.BitsPerSample + 7) / 8);
            format.AverageBytesPerSecond = format.BlockAlign * format.SamplesPerSecond;

            bufferLength = format.AverageBytesPerSecond * BufferSeconds;
            CaptureBufferDescription desciption = new CaptureBufferDescription();
            desciption.Format = format;
            desciption.BufferBytes = bufferLength;

            capture = new Capture(device.Id);
            buffer = new CaptureBuffer(desciption, capture);

            int waitHandleCount = BufferSeconds * NotifyPointsInSecond;
            BufferPositionNotify[] positions = new BufferPositionNotify[waitHandleCount];
            for (int i = 0; i < waitHandleCount; i++)
            {
                BufferPositionNotify position = new BufferPositionNotify();
                position.Offset = (i + 1) * bufferLength / positions.Length - 1;
                position.EventNotifyHandle = positionEventHandle.DangerousGetHandle();
                positions[i] = position;
            }

            notify = new Notify(buffer);
            notify.SetNotificationPositions(positions);

            terminated.Reset();
            thread = new Thread(new ThreadStart(ThreadLoop));
            thread.Name = "Sound capture";
            thread.Start();
        }

        private void ThreadLoop()
        {
            buffer.Start(true);
            try
            {
                int nextCapturePosition = 0;
                WaitHandle[] handles = new WaitHandle[] { terminated, positionEvent };
                while (WaitHandle.WaitAny(handles) > 0)
                {
                    int capturePosition, readPosition;
                    buffer.GetCurrentPosition(out capturePosition, out readPosition);

                    int lockSize = readPosition - nextCapturePosition;
                    if (lockSize < 0) lockSize += bufferLength;
                    if((lockSize & 1) != 0) lockSize--;

                    int itemsCount = lockSize >> 1;

                    short[] data = (short[])buffer.Read(nextCapturePosition, typeof(short), LockFlag.None, itemsCount);
                    ProcessData(data);
                    nextCapturePosition = (nextCapturePosition + lockSize) % bufferLength;
                }
            }
            finally
            {
                buffer.Stop();
            }
        }

        /// <summary>
        /// Processes the captured data.
        /// </summary>
        /// <param name="data">Captured data</param>
        protected abstract void ProcessData(short[] data);

        /// <summary>
        /// Stops capture process.
        /// </summary>
        public void Stop()
        {
            if (isCapturing)
            {
                isCapturing = false;

                terminated.Set();
                thread.Join();

                notify.Dispose();
                buffer.Dispose();
                capture.Dispose();
            }
        }

        void IDisposable.Dispose()
        {
            Dispose(true);
        }

        ~SoundCaptureBase()
        {
            Dispose(false);
        }

        private void Dispose(bool disposing)
        {
            if (disposed) return;

            disposed = true;
            GC.SuppressFinalize(this);
            if (IsCapturing) Stop();
            positionEventHandle.Dispose();
            positionEvent.Close();
            terminated.Close();            
        }
    }
}
